### DISCUSSION TITLE
How to render all versions of a file?

### DISCUSSION THREAD

**Discussion by Mark2Mark** [coldtype/coldtype#158] [Q&A]
When using the `VERSIONS`/`__VERSION__` macro one can switch those in the viewer with <kbd>shift</kbd>+<kbd>v</kbd>, and then render (<kbd>a</kbd>/<kbd>r</kbd>) the current version. 

Is there a way to render all versions at once?

Currently also, when rendering one version, switching to another one and rendering that, it overrides the previous rendering.

I found that the option `-rar` does go through all versions, create the frames and the video. But that still overrides each previous pass, when it renders the next version. So I think the version-number might need to be added to the rendering or sth like that?

**Comment by stenson** [comment]
The `KeyboardShortcut.RenderAllAndRelease` (shift+cmd+a) (or `-rar` from the command line as you pointed out) should work properly, I'm wondering if you're not including the `ƒVERSION` macro in the `@animation` name as demonstrated in https://github.com/coldtype/coldtype/blob/main/examples/animations/versioned.py? Definitely a little obscure, but lets you rename the function to prevent overwriting like you're describing. That could be done automatically, but I think it's helpful to have the VERSION macros not enforce the idea that the render target necessarily changes whenever versions are enabled.

**Reply by Mark2Mark** [reply]
Oh, I see. I saw that `versioned_ƒVERSION` thingy, but didn’t really expect it to be required.
I found out about the versions from ArrowType’s Shantell Repo, where you use it. But the files that use the `VERSIONS = {...` dict, don’t have the `versioned_ƒVERSION`. 

That got me confused about the relevance of it. `ƒ` is also a little hard to type. Did you consider giving it a different name?

But it works indeed. Thank you! I totally love this feature!

I also found out by trial and error that the versions dict **must** end with `} #/VERSIONS`. Also here I thought that is not relevant, until I found it in several places. My black formatter adds a space after the `#`, so I need to disable it for this. In my files it looks like this now :D 
```
# fmt: off
} #/VERSIONS
# fmt: on
```

Such things are really really cool, but it’s still a lot of poking around to find them :)
Sorry for asking so many silly questions.

**Reply by stenson** [reply]
Definitely not a silly question — that said, I haven't spent too much time on documenting the VERSIONS stuff, since it was an idea first implemented for Shantell Sans (though I have used it a couple times on work projects). Rather than use ƒVERSION, the Shantell Sans code uses something like `name="kinetic4_" + __VERSION__["script"]`, which accomplishes the same thing by completely overriding the name of the renderable/animation (which is otherwise taken directly from the function name). So it sounds like that might be a better solution for your case?

The `#/VERSIONS` bit I could definitely tweak to allow an optional space, but yes it is necessary because it's actually a pre-runtime macro, meaning there's code that reads from `#VERSIONS` to `#/VERSIONS` (ala a closing tag in html), and reads that bit as a literal, which is why you can't have variables in it.

There is another example in the coldtype repo that demonstrates a "sidecar" versioning concept, which lets you put versions data in a standard python file, where you can use any variables you'd like. The file has to be named exactly the same as the main file, just  with "_versions" at the end of it. Here are the two files that make up that example: https://github.com/coldtype/coldtype/blob/main/examples/animations/versioned_with_sidecar.py and https://github.com/coldtype/coldtype/blob/main/examples/animations/versioned_with_sidecar_versions.py — so in that example, it's reading file names from disk, rather than using any literal data.

All-in-all, I think the `VERSIONS` idea is still very much experimental, though also very useful, as it lets you quickly re-use code that would otherwise be one-off. So if people like it, I could definitely "cement" it a little more in coldtype.

**Reply by Mark2Mark** [reply]
Interesting that it’s so new. For me that is already one of may all time fav features. I have so many animations that use different aspect ratios or color schemes, and now I can merge all those files into one (instead of having foo_1_color_1_square.py, foo_1_color_1_portrait.py, foo_1_color_2_square.py, foo_1_color_2_portrait.py, and so on 😆 )

So one big love (not just like) from me for the versions feature!

> So it sounds like that might be a better solution for your case?

No, it already works well with the `ƒVERSION` :)

I will also have a look into the sidecar. 
One note: The comment says `defined in *._versions.py versions of this file`, but the file is `versioned_with_sidecar_versions.py` (without period). As you say it has to be exact, which one is it?

**Comment by Mark2Mark** [comment]
One other question about the VERSIONS dictionary: I cannot add variables inside this dictionary, or I will get a not defined error, even though I declare the variable right before.

But there is some kind of wizzardry happening with this macro (as observed before in this thread), that puzzles me about that question.

```py
foo = 123
VERSIONS = {
    "a" = dict(my_number=foo) # won’t work
} #/VERSIONS
```
